File: frontend/src/types/team.ts
/**
 * Type definitions for team/squad data structures
 */

export interface PlayerDetail {
  id: number
  name: string
  position: "GK" | "DEF" | "MID" | "FWD"
  team: string
  team_short_name?: string
  price: number // Price in millions (e.g., 10.5 = £10.5M)
  fpl_code?: number | null
  team_fpl_code?: number | null // FPL team code for shirt images
  status: "a" | "i" | "s" | "u" | "d" // a=available, i=injured, s=suspended, u=unavailable, d=doubtful
  is_starting?: boolean
  is_captain?: boolean
  is_vice_captain?: boolean
  multiplier?: number
  gw_points_raw?: number
  gw_points?: number
  total_points: number
  goals_scored: number
  assists: number
  clean_sheets: number
  yellow_cards?: number // Yellow cards
  red_cards?: number // Red cards
  // Fixture information
  next_fixture_opponent?: string // Opponent team name
  next_fixture_opponent_short?: string // Opponent team short code (e.g., "LIV", "MCI")
  next_fixture_home_away?: "H" | "A" // Home or Away
  next_fixture_difficulty?: number // 1-5 difficulty rating
}

export interface TeamEvaluationResponse {
  season: string
  gameweek: number | null
  overall_points?: number | null
  gw_rank?: number | null
  transfers?: number | null
  active_chip?: string | null
  total_points: number
  expected_points: number
  xg_score?: number
  risk_score: number
  fixture_difficulty: number
  squad_value: number // Total squad value in millions
  bank: number // Remaining bank in millions
  players: PlayerDetail[]
  captain_id: number | null
  vice_captain_id: number | null
}

export interface Formation {
  name: string
  def: number
  mid: number
  fwd: number
}

export const FORMATIONS: Formation[] = [
  { name: "4-4-2", def: 4, mid: 4, fwd: 2 },
  { name: "4-3-3", def: 4, mid: 3, fwd: 3 },
  { name: "3-5-2", def: 3, mid: 5, fwd: 2 },
  { name: "3-4-3", def: 3, mid: 4, fwd: 3 },
  { name: "4-5-1", def: 4, mid: 5, fwd: 1 },
  { name: "5-4-1", def: 5, mid: 4, fwd: 1 },
  { name: "5-3-2", def: 5, mid: 3, fwd: 2 },
]

File: backend/app/api/v1/schemas/team.py
"""Schemas for team-related endpoints."""
from __future__ import annotations
from typing import Optional, List
from pydantic import BaseModel, Field


class PlayerBase(BaseModel):
    """Base player schema."""
    id: int
    name: str
    position: str
    team: str
    price: float


class PlayerDetail(PlayerBase):
    """Detailed player schema."""
    fpl_code: Optional[int] = None
    team_fpl_code: Optional[int] = None  # FPL team code for shirt images
    team_short_name: Optional[str] = None
    status: str = "a"
    is_starting: Optional[bool] = None
    is_captain: Optional[bool] = None
    is_vice_captain: Optional[bool] = None
    multiplier: Optional[int] = None
    gw_points_raw: Optional[float] = None
    gw_points: Optional[float] = None
    total_points: float = 0.0
    goals_scored: int = 0
    assists: int = 0
    clean_sheets: int = 0


class TeamEvaluationRequest(BaseModel):
    """Request schema for team evaluation."""
    season: str = Field(..., description="Season identifier, e.g., '2024-25'")
    team_id: Optional[int] = Field(None, description="FPL team ID")
    squad_json: Optional[dict] = Field(None, description="Custom squad JSON payload")
    gameweek: Optional[int] = Field(None, description="Target gameweek for evaluation")


class SquadMember(BaseModel):
    """Squad member schema."""
    player_id: int
    position: str
    is_starting_xi: bool = True
    is_captain: bool = False
    is_vice_captain: bool = False


class TeamEvaluationResponse(BaseModel):
    """Response schema for team evaluation."""
    season: str
    gameweek: Optional[int]
    overall_points: Optional[float] = None
    gw_rank: Optional[int] = None
    transfers: Optional[int] = None
    active_chip: Optional[str] = None
    total_points: float
    expected_points: float
    risk_score: float = Field(..., ge=0.0, le=1.0, description="Risk score 0-1")
    fixture_difficulty: float = Field(..., ge=1.0, le=5.0, description="Fixture difficulty 1-5")
    squad_value: float
    bank: float
    players: List[PlayerDetail]
    captain_id: Optional[int] = None
    vice_captain_id: Optional[int] = None


class XGScoreRequest(BaseModel):
    """Request schema for XG Score calculation."""
    season: str
    squad: List[SquadMember]
    gameweek: Optional[int] = None


class XGScoreResponse(BaseModel):
    """Response schema for XG Score."""
    xg_score: float = Field(..., description="Computed XG Score")
    components: dict = Field(..., description="Score component breakdown")
    ml_prediction_score: float
    fixture_adjusted_score: float
    captaincy_bonus: float
    risk_adjusted_score: float
    monte_carlo_mean: float
    monte_carlo_std: float
    percentile_rank: float = Field(..., ge=0.0, le=100.0)

